File: cpp/src/0326.cpp
/*
[326. Power of Three (Easy)](https://leetcode.com/problems/power-of-three/)

@topic: Math
@topic: Recursion

Constraints:
  - `-2^31 <= n <= 2^31 - 1`
*/

#include <unordered_set>

class Solution {
   public:
    bool isPowerOfThree(int n) {
        return n > 0 && 1162261467 % n == 0;
    }
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    Solution o;

    vector<tuple<int, bool>> CASES = {
        {27, true},
        {0, false},
        {9, true},
    };

    for (auto& [n, excepted] : CASES) {
        assert(o.isPowerOfThree(n) == excepted);
    }

    auto start = system_clock::now();
    for (auto& [n, _] : CASES) {
        for (int i = 0; i < 100000; ++i) {
            o.isPowerOfThree(n);
        }
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}

File: python/src/0326.py
class Solution:
    def isPowerOfThree(self, n: int) -> bool:
        return n > 0 and 3**19 % n == 0


if __name__ == '__main__':
    import unittest
    from timeit import timeit

    CASES = (
        (0, False),
        (1, True),
        (2, False),
        (3, True),
        (4, False),
        (5, False),
        (6, False),
        (7, False),
        (8, False),
        (9, True),
        (10, False),
        (27, True),
        (45, False),
        (81, True),
        (243, True),
        (729, True),
        (2187, True),
        (6561, True),
        (19683, True),
    )

    class Test(unittest.TestCase):
        def test(self):
            isPowerOfThree = Solution().isPowerOfThree
            for n, excepted in CASES:
                actual = isPowerOfThree(n)
                self.assertEqual(actual, excepted)

    unittest.main()

    isPowerOfThree = Solution().isPowerOfThree
    for n, _ in CASES:
        print(timeit(lambda: isPowerOfThree(n), number=10000))
